src/zks_wallet/wallet.rs

```diff
@@ -432,80 +430,126 @@ where
         })
     }
 
-    pub async fn send<T>(
-        &self,
-        contract_address: Address,
-        function_signature: &str,
-        function_parameters: Option<T>,
-    ) -> Result<(Vec<Token>, H256), ZKSWalletError<M, D>>
+    pub async fn withdraw(&self, amount: U256) -> Result<TransactionReceipt, ZKSWalletError<M, D>>
     where
         M: ZKSProvider,
-        T: Tokenizable + Debug + Clone,
     {
         let era_provider = match &self.era_provider {
             Some(era_provider) => era_provider,
             None => return Err(ZKSWalletError::CustomError("no era provider".to_owned())),
         };
 
-        // Note: We couldn't implement ZKSWalletError::LexerError because ethers-rs's LexerError is not exposed.
-        let function = HumanReadableParser::parse_function(function_signature)
-            .map_err(|e| ZKSWalletError::CustomError(e.to_string()))?;
-
-        let mut send_request = Eip712TransactionRequest::new()
-            .r#type(EIP712_TX_TYPE)
-            .from(self.address())
-            .to(contract_address)
-            .chain_id(ERA_CHAIN_ID)
-            .nonce(
-                era_provider
-                    .get_transaction_count(self.address(), None)
-                    .await?,
+        let contract_address = Address::from_str(CONTRACTS_L2_ETH_TOKEN_ADDR).unwrap();
+        let function_signature = "function withdraw(address _l1Receiver) external payable override";
+        let response: (Vec<Token>, H256) = era_provider
+            .send_eip712(
+                &self.wallet,
+                contract_address,
+                function_signature,
+                Some(self.wallet.address()),
+                Some(Overrides {
+                    value: Some(amount),
+                }),
             )
-            .gas_price(era_provider.get_gas_price().await?)
-            .max_fee_per_gas(era_provider.get_gas_price().await?)
-            .data(match function_parameters {
-                Some(parameters) => function
-                    .encode_input(&parameters.into_tokens())
-                    .map_err(|e| ZKSWalletError::CustomError(e.to_string()))?,
-                None => function.short_signature().into(),
-            });
-
-        let fee = era_provider.estimate_fee(send_request.clone()).await?;
-        send_request = send_request
-            .max_priority_fee_per_gas(fee.max_priority_fee_per_gas)
-            .max_fee_per_gas(fee.max_fee_per_gas)
-            .gas_limit(fee.gas_limit);
+            .await?;
 
-        let signable_data: Eip712Transaction = send_request.clone().try_into()?;
-        let signature: Signature = self.wallet.sign_typed_data(&signable_data).await?;
-        send_request =
-            send_request.custom_data(Eip712Meta::new().custom_signature(signature.to_vec()));
+        Ok(era_provider
+            .get_transaction_receipt(response.1)
+            .await
+            .unwrap()
```

Comment:
should we return an error instead of panicking here?

Reply:
Already fixed!